### PR TITLE
Fixes Loss for TransfoXL when using Trainer API v2

### PR DESCRIPTION
**Pull request by LysandreJik**
Reopens https://github.com/huggingface/transformers/pull/14568#issue-1066398357 with backward-compatible mechanisms

### CODE REVIEW

File: src/transformers/models/transfo_xl/modeling_transfo_xl.py [hunk 1]
```diff
@@ -692,13 +693,16 @@ class TransfoXLLMHeadModelOutput(ModelOutput):
...
+        loss (`torch.FloatTensor` of shape *()*, *optional*, returned when `labels` is provided)
+            Reduced language modeling loss.
     """
 
     losses: Optional[torch.FloatTensor] = None
     prediction_scores: torch.FloatTensor = None
     mems: List[torch.FloatTensor] = None
     hidden_states: Optional[Tuple[torch.FloatTensor]] = None
     attentions: Optional[Tuple[torch.FloatTensor]] = None
+    loss: Optional[torch.FloatTensor] = None
```
Comment (sgugger) [inline src/transformers/models/transfo_xl/modeling_transfo_xl.py:705]: This needs to be first to work with the `Trainer` and `return_dict=False`, but maybe we don't care about this use case.

**Review by sgugger — APPROVED** [review]
Thanks for reviving this!

File: src/transformers/models/transfo_xl/modeling_transfo_xl.py [hunk 2]
```diff
@@ -1095,17 +1107,38 @@ def forward(
...
+        else:
+            losses, loss = None, None
 
         if not return_dict:
-            output = (prediction_scores,) + transformer_outputs[1:]
-            return ((loss,) + output) if loss is not None else output
+            if self.trainer_compatible:
+                output = (prediction_scores, losses) if losses is not None else (prediction_scores,)
+                output += transformer_outputs[1:]
+                return ((loss,) + output) if loss is not None else output
```
Comment (LysandreJik) [inline src/transformers/models/transfo_xl/modeling_transfo_xl.py:1132]: @sgugger here we place it first in the return, shouldn't that be sufficient?

Comment (sgugger) [reply src/transformers/models/transfo_xl/modeling_transfo_xl.py:1132]: Yes, it should, you're right!

**Review by patrickvonplaten — APPROVED** [review]
